### PR TITLE
Using Kahan's algorithm to compute AvgFunction and SumFunction

### PR DESCRIPTION
**Pull request by craigtaverner**
I cherry-picked the original commit from 3.0 to 2.3. The original commit message claims:

> Refactored Kahan's summation method into a separate class ; using it for SUM as well as AVG
> 
>  \* Updated docs for AVG/SUM/KahanSum
> - Made SUM and AVG use gradually type-widening algorithms
> - Using own impl. of Math#addExact to fit JDK <= 7

### CODE REVIEW

**Comment by craigtaverner** [comment]
Turns out the fix causes sum() to return int instead of long if int, short or byte are passed in. Before merging we should confirm the impact of this behavior change, or fix it.

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/pipes/aggregation/SumFunction.scala [hunk 1]
```diff
@@ -31,11 +31,12 @@ class SumFunction(val value: Expression)
 
   def name = "SUM"
 
-  var result: Any = 0
+  private var sum: OverflowAwareSum[_] = OverflowAwareSum(0)
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/pipes/aggregation/SumFunction.scala]: change this to `OverflowAwareSum(0L)` and revert the change in `SumFunctionTest`

**Comment by davidegrohmann** [comment]
@craigtaverner why not backporting this back to 2.2 as well?
